File: harness-sdk/opencode_stream.py
"""Parse OpenCode run --format json output into structured harness events."""

from __future__ import annotations

import json
from typing import Any, Generator, Iterator

from claude_stream import _emit_image_events


def parse_opencode_stream(lines: Iterator[str]) -> Generator[dict[str, Any], None, None]:
    parser = new_opencode_stream_parser()
    for line in lines:
        line = line.strip()
        if not line:
            continue
        try:
            obj = json.loads(line)
        except json.JSONDecodeError:
            continue
        yield from parser.handle(obj)


def new_opencode_stream_parser() -> "_OpenCodeStreamParser":
    return _OpenCodeStreamParser()


class _OpenCodeStreamParser:
    def __init__(self) -> None:
        self.seen_tool_starts: set[str] = set()
        self.seen_tool_ends: set[str] = set()
        self.seen_tool_results: set[str] = set()

    def handle(self, obj: dict[str, Any]) -> Generator[dict[str, Any], None, None]:
        sid = obj.get("sessionID") or ""
        if sid:
            yield {"type": "session_id", "sessionId": sid}

        t = obj.get("type")
        if t == "text":
            part = obj.get("part") or {}
            if part.get("type") == "text":
                text = part.get("text") or ""
                if text:
                    yield {"type": "text", "content": text}
            return

        if t != "tool_use":
            return

        part = obj.get("part") or {}
        state = part.get("state") or {}
        if state.get("status") != "completed":
            return

        tool_id = part.get("callID") or part.get("id") or ""
        tool_name = part.get("tool") or ""
        tool_input = state.get("input") or {}
        tool_output = state.get("output")
        metadata = state.get("metadata") or {}

        if not tool_id:
            return

        if tool_id not in self.seen_tool_starts:
            self.seen_tool_starts.add(tool_id)
            yield {"type": "tool_call_start", "toolCallId": tool_id, "toolName": tool_name}

        args_json = json.dumps(tool_input)
        yield {"type": "tool_call_args", "toolCallId": tool_id, "toolArgs": args_json}

        if tool_id not in self.seen_tool_ends:
            self.seen_tool_ends.add(tool_id)
            yield {
                "type": "tool_call_end",
                "toolCallId": tool_id,
                "toolName": tool_name,
                "toolArgs": args_json,
            }

        if tool_id not in self.seen_tool_results:
            self.seen_tool_results.add(tool_id)
            result_payload: Any = tool_output
            if isinstance(tool_output, str):
                try:
                    result_payload = json.loads(tool_output)
                except json.JSONDecodeError:
                    result_payload = tool_output
            elif metadata:
                result_payload = {"output": tool_output, "metadata": metadata}

            content = result_payload if isinstance(result_payload, str) else json.dumps(result_payload)
            yield {"type": "tool_call_result", "toolCallId": tool_id, "content": content}
            yield from _emit_image_events(result_payload)
            yield from _emit_image_events(metadata)
            yield from _emit_image_events(state)

File: internal/agent/opencode_stream.go
// Copyright (c) Mehmet Bektas <mbektasgh@outlook.com>

package agent

import (
	"encoding/json"
)

type openCodeStreamParser struct {
	seenToolStarts  map[string]bool
	seenToolEnds    map[string]bool
	seenToolResults map[string]bool
}

func newOpenCodeStreamParser() *openCodeStreamParser {
	return &openCodeStreamParser{
		seenToolStarts:  map[string]bool{},
		seenToolEnds:    map[string]bool{},
		seenToolResults: map[string]bool{},
	}
}

func (p *openCodeStreamParser) handleLine(line []byte, events chan<- Event) {
	var obj struct {
		Type string `json:"type"`
		Part struct {
			Type   string `json:"type"`
			Text   string `json:"text"`
			CallID string `json:"callID"`
			ID     string `json:"id"`
			Tool   string `json:"tool"`
			State  struct {
				Status string          `json:"status"`
				Input  json.RawMessage `json:"input"`
				Output json.RawMessage `json:"output"`
				Metadata json.RawMessage `json:"metadata"`
			} `json:"state"`
		} `json:"part"`
	}
	if err := json.Unmarshal(line, &obj); err != nil {
		return
	}

	if obj.Type == "text" && obj.Part.Type == "text" && obj.Part.Text != "" {
		events <- Event{Type: EventText, Content: obj.Part.Text}
		return
	}
	if obj.Type != "tool_use" {
		return
	}
	if obj.Part.State.Status != "completed" {
		return
	}

	toolID := obj.Part.CallID
	if toolID == "" {
		toolID = obj.Part.ID
	}
	if toolID == "" {
		return
	}

	toolName := obj.Part.Tool
	argsJSON, _ := json.Marshal(obj.Part.State.Input)
	if string(argsJSON) == "null" {
		argsJSON = []byte("{}")
	}

	if !p.seenToolStarts[toolID] {
		p.seenToolStarts[toolID] = true
		events <- Event{
			Type:       EventToolCallStart,
			ToolCallID: toolID,
			ToolName:   toolName,
		}
	}
	events <- Event{
		Type:       EventToolCallArgs,
		ToolCallID: toolID,
		ToolArgs:   string(argsJSON),
	}
	if !p.seenToolEnds[toolID] {
		p.seenToolEnds[toolID] = true
		events <- Event{
			Type:       EventToolCallEnd,
			ToolCallID: toolID,
			ToolName:   toolName,
			ToolArgs:   string(argsJSON),
		}
	}
	if p.seenToolResults[toolID] {
		return
	}
	p.seenToolResults[toolID] = true

	resultPayload := obj.Part.State.Output
	if len(resultPayload) > 0 && resultPayload[0] == '"' {
		var decoded string
		if json.Unmarshal(resultPayload, &decoded) == nil {
			var parsed any
			if json.Unmarshal([]byte(decoded), &parsed) == nil {
				resultPayload, _ = json.Marshal(parsed)
			} else {
				resultPayload = []byte(decoded)
			}
		}
	} else if len(obj.Part.State.Metadata) > 0 && string(obj.Part.State.Metadata) != "null" {
		resultPayload, _ = json.Marshal(map[string]any{
			"output":   rawJSONToAny(obj.Part.State.Output),
			"metadata": rawJSONToAny(obj.Part.State.Metadata),
		})
	}

	events <- Event{
		Type:       EventToolCallResult,
		ToolCallID: toolID,
		Content:    string(resultPayload),
	}
	emitImageEvents(resultPayload, "", events)
	emitImageEvents(obj.Part.State.Metadata, "", events)
	if stateRaw, err := json.Marshal(obj.Part.State); err == nil {
		emitImageEvents(stateRaw, "", events)
	}
}

func rawJSONToAny(raw json.RawMessage) any {
	if len(raw) == 0 || string(raw) == "null" {
		return nil
	}
	var v any
	if json.Unmarshal(raw, &v) == nil {
		return v
	}
	return string(raw)
}
